aiohttp/cookiejar.py

```diff
@@ -280,10 +280,23 @@ def filter_cookies(
             if is_not_secure and cookie["secure"]:
                 continue
 
-            # It's critical we use the Morsel so the coded_value
-            # (based on cookie version) is preserved
-            mrsl_val = cast("Morsel[str]", cookie.get(cookie.key, Morsel()))
-            mrsl_val.set(cookie.key, cookie.value, cookie.coded_value)
+            # we need to check, if the found cookie is a better match then the
+            # previously found
+            if name in filtered and len(cookie["path"]) < len(filtered[name]["path"]):
+                continue
+
+            filtered[name] = cookie
+
+        # It's critical we use the Morsel so the coded_value
+        # (based on cookie version) is preserved
+        for name in filtered:
+            filtered_cookie = filtered[name]
+            mrsl_val = cast(
+                "Morsel[str]", filtered_cookie.get(filtered_cookie.key, Morsel())
+            )
+            mrsl_val.set(
+                filtered_cookie.key, filtered_cookie.value, filtered_cookie.coded_value
+            )
```

Comment:
In my first comment. I meant that this whole section could also be reverted, we only need 2 lines of code added. Unless I'm misunderstanding something.

Reply:
Right, in this case, I think it might make more sense to just sort the cookies before iterating through them (as suggested by the spec, and done in the upstream CookieJar: https://github.com/python/cpython/blob/a0773b89dfe5cd2190d539905dd89e7f6455668e/Lib/http/cookiejar.py#L1307).

Reply:
Then we won't need to check the cookies here. If we do reverse sorting, then we don't even need a check here at all, the last (longest path) cookie will be the one which is set.

Reply:
Again I fully agree with your suggestion (to sort the "cookie" storage) - but I have to admit, with my current (python) knowledge I am not able to implement such a functionality... If you have an idea, how to do that - please do so and just cancel this PR.

Reply:
I would've thought copy/paste the sort function onto `iter(self)` would suffice.
Probably just: `sorted(iter(self), key=lambda c: len(c["path"]))`

I noticed there is also a tie-breaker in the spec too:
```
       *  Among cookies that have equal-length path fields, cookies with
          earlier creation-times are listed before cookies with later
          creation-times.
```
But, I don't think we store creation time currently, so will have to ignore for now.

Reply:
Actually, the `iter()` shouldn't even be needed.

Reply:
I tried to include the line `sorted(self, key=lambda c: len(c["path"]))` after the `self._do_expiration()` - but the content of the self._cookies does not change it's order... [inspecting in the debugger]

The result/return-type of the sorted method is IMHO also different type then the Jar itself... I tried also sorting something else then 'self' (self._cookies or self.cookies.values()) - but all that was not successful... please don't laugh too loud (it might be totally stupid what I am trying to do based on my java knowledge)

Reply:
_cookies is a dict (nested dicts as well...), so it would be tricky to resort it in-place. That function just returns a sorted list of all the cookies. Just use the return result in the `for` loop.

Reply:
When you iterate over `self` (which the sorted() function is doing), it just calls the `__iter__()` method:
https://github.com/aio-libs/aiohttp/blob/4639b36852f3b4939e61311d7863f1d7450d500d/aiohttp/cookiejar.py#L141-L142

As you can see from that (assuming you can follow the `yield` syntax), it basically iterates over every cookie in the nested dicts. So, that's why you see a difference in types.

Reply:
should I (or have to) squash now all commits? - since "now" with the latest commit I guess we are done?